lib/Dialect/LLHD/Simulator/State.cpp

```diff
@@ -0,0 +1,257 @@
+//===- State.cpp - LLHD simulator state -------------------------*- C++ -*-===//
+//
+// This file implements the constructs used to keep track of the simulation
+// state in the LLHD simulator.
+//
+//===----------------------------------------------------------------------===//
+
+#include "State.h"
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <iomanip>
+#include <sstream>
+#include <string>
+
+using namespace llvm;
+using namespace mlir;
+using namespace llhd::sim;
+
+//===----------------------------------------------------------------------===//
+// Time
+//===----------------------------------------------------------------------===//
+
+bool Time::operator<(const Time &rhs) const {
+  if (time < rhs.time)
+    return true;
+  if (time == rhs.time && delta < rhs.delta)
+    return true;
+  if (time == rhs.time && delta == rhs.delta && eps < rhs.eps)
+    return true;
+  return false;
+}
+
+bool Time::operator==(const Time &rhs) const {
+  return (time == rhs.time && delta == rhs.delta && eps == rhs.eps);
+}
+
+Time Time::operator+(const Time &rhs) const {
+  return Time(time + rhs.time, delta + rhs.delta, eps + rhs.eps);
+}
+
+bool Time::isZero() { return (time == 0 && delta == 0 && eps == 0); }
+
+std::string Time::dump() {
+  std::stringstream dumpStr;
+  dumpStr << time << "ns " << delta << "d " << eps << "e";
+  return dumpStr.str();
+}
+
+//===----------------------------------------------------------------------===//
+// Signal
+//===----------------------------------------------------------------------===//
+
+Signal::Signal(std::string name, std::string owner)
+    : name(name), owner(owner), size(0), detail({nullptr, 0}) {}
+
+Signal::Signal(std::string name, std::string owner, uint8_t *value,
+               uint64_t size)
+    : name(name), owner(owner), size(size), detail({value, 0}) {}
+
+Signal::Signal(int origin, uint8_t *value, uint64_t size, uint64_t offset)
+    : origin(origin), size(size), detail({value, offset}) {}
+
+bool Signal::operator==(const Signal &rhs) const {
+  if (owner != rhs.owner || name != rhs.name || size != rhs.size)
+    return false;
+  for (uint64_t i = 0; i < size; ++i) {
+    if (detail.value[i] != rhs.detail.value[i])
+      return false;
+  }
+  return true;
+}
+
+bool Signal::operator<(const Signal &rhs) const {
+  if (owner < rhs.owner)
+    return true;
+  if (owner == rhs.owner && name < rhs.name)
+    return true;
+  return false;
+}
+
+std::string Signal::dump() {
+  std::stringstream ss;
```

Comment:
I'd recommend using llvm/Support/Format.h instead of sstream and iomanip.  It lets you do things like:

os << "mynumber: " << llvm::format("%4.5f", 1234.412) << '\n';

Reply:
That's way cleaner, thank you!